### ISSUE TITLE
aggregation for all metrics

### ISSUE THREAD

**Issue by randallt** [graphite-ng/carbon-relay-ng#238]
I can't seem to get an aggregation section defined to simply aggregate ALL incoming metrics to 30-sec boundaries and pass them through with metric names unchanged. I see no aggregation happening at all.

My aggregation section currently looks like:

```
[[aggregation]]
# aggregate timer metrics with averages
function = 'avg'
regex = '(.*)'
format = '$1'
interval = 30
wait = 90
cache = true
```

How can I simply do this for all metrics?

FYI, this worked easily in the python carbon-aggregator, with this in aggregation-rules.conf:
```
<metric> (30) = avg <<metric>>
```

**Comment by randallt** [comment]
Is the issue that I AM aggregating these metrics, but also that the non-aggregated metrics are passing through to the destinations? How can I configure this so that ONLY the aggregated metrics are sent to the destinations (so that I only have one data point for each metric every 30 sec)?

**Comment by Dieterbe** [comment]
with the current implementation, both the source metrics as well as the aggregates go into the routing pipeline. if keeping the names the same, there is thus no way to match the aggregates but not the input metrics.
it should be fairly trivial to add a flag to aggregators so you can set them to discard the inputs or not. but before I do I'ld like to understand the use case better. e.g. can't you have the source report the metrics in 30s intervals to begin with?
also, do you know how it works with carbon-aggregator, does it just automatically discard the input metrics if it detects the output name is the same? because if the names do differ, the aggregator should also send original + aggregated metrics, right?

**Comment by randallt** [comment]
I cannot change the frequency of the source metrics. If I could, I wouldn’t need the aggregator. I do not know how carbon-aggregator does it—can look into that later. I think a flag to drop the source metrics is a fine way to enable this function.

The use case is that I am relating to a SaaS metric collector, where I pay per metric.  So I need a way to dial down the frequency.


Randall Theobald
Software Sr Principal Engineer
SecureWorks | Performance Engineering
rtheobald@secureworks.com
(Sent from mobile)


On November 29, 2017 at 11:42:23 AM PST, Dieter Plaetinck <notifications@github.com> wrote:

with the current implementation, both the source metrics as well as the aggregates go into the routing pipeline. if keeping the names the same, there is thus no way to match the aggregates but not the input metrics.
it should be fairly trivial to add a flag to aggregators so you can set them to discard the inputs or not. but before I do I'ld like to understand the use case better. e.g. can't you have the source report the metrics in 30s intervals to begin with?
also, do you know how it works with carbon-aggregator, does it just automatically discard the input metrics if it detects the output name is the same? because if the names do differ, the aggregator should also send original + aggregated metrics, right?

**Comment by randallt** [comment]
The Python carbon-aggregator does like you guessed--it only forwards on metrics that don't match an aggregated metric. That is, aggregated metrics override source metrics:

From carbon.aggregator.receiver.py:
```
if settings['FORWARD_ALL'] and metric not in aggregate_metrics:
    events.metricGenerated(metric, datapoint)
```

Might also be worth a global flag, like the FORWARD_ALL in carbon-aggregator, which can be set to false.

**Comment by randallt** [comment]
Is there a way I can mix-and-match an aggregator and rewrite rules to possibly achieve this without a new feature? If not, could it be done with multiple carbon-relay-ng processes? We need this quickly for a PoC.

**Comment by randallt** [comment]
I have this functioning using this config below. Basically all incoming metrics get rewritten to have an '@' prefix. The aggregation works on these, but outputs to the original metric name. The route only matches metrics that don't have an '@'.
```
[[rewriter]]
# rewrite all instances to have @ prefix (for aggregation)
old = '/^/'
new = '@'
max = -1

# Aggregators
[[aggregation]]
# aggregate rewritten metrics with averages, writing back to original metric name
function = 'avg'
regex = '^@(.*)'
format = '$1'
interval = 30
wait = 60
cache = true

# Routes
[[route]]
# route only metrics that don't start with '@'
key = 'carbon-default'
type = 'sendAllMatch'
regex = '^[^@]'
destinations = [
  'XXXXX'
]
```

**Comment by Dieterbe** [comment]
Very clever. Should work fine and not pile up data. I still want to add the config flag but have not gotten around to it

**Comment by Dieterbe** [comment]
dropRaw was implemented last year, via #268
